matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/rendezvous/channels/ECDHRendezvousChannel.kt

```diff
@@ -0,0 +1,219 @@
+        @Json val iv: String? = null
+)
+
+private val TAG = LoggerTag(ECDHRendezvousChannel::class.java.simpleName, LoggerTag.RENDEZVOUS).value
+
+fun getDecimalCodeRepresentation(byteArray: ByteArray): String {
+    val b0 = byteArray[0].toUnsignedInt() // need unsigned byte
+    val b1 = byteArray[1].toUnsignedInt() // need unsigned byte
+    val b2 = byteArray[2].toUnsignedInt() // need unsigned byte
+    val b3 = byteArray[3].toUnsignedInt() // need unsigned byte
+    val b4 = byteArray[4].toUnsignedInt() // need unsigned byte
+    // (B0 << 5 | B1 >> 3) + 1000
+    val first = (b0.shl(5) or b1.shr(3)) + 1000
+    // ((B1 & 0x7) << 10 | B2 << 2 | B3 >> 6) + 1000
+    val second = ((b1 and 0x7).shl(10) or b2.shl(2) or b3.shr(6)) + 1000
+    // ((B3 & 0x3f) << 7 | B4 >> 1) + 1000
+    val third = ((b3 and 0x3f).shl(7) or b4.shr(1)) + 1000
+    return "$first-$second-$third"
+}
+
+const val ALGORITHM_SPEC = "AES/GCM/NoPadding"
+const val KEY_SPEC = "AES"
+
+/**
+ *  Implements X25519 ECDH key agreement and AES-256-GCM encryption channel as per MSC3903:
+ *  https://github.com/matrix-org/matrix-spec-proposals/pull/3903
+ */
+class ECDHRendezvousChannel(override var transport: RendezvousTransport, theirPublicKeyBase64: String?) : RendezvousChannel {
+    private var olmSAS: OlmSAS?
+    private val ourPublicKey: ByteArray
+    private val ecdhAdapter = MatrixJsonParser.getMoshi().adapter(ECDHPayload::class.java)
+    private var theirPublicKey: ByteArray? = null
+    private var aesKey: ByteArray? = null
+
+    init {
+        theirPublicKeyBase64 ?.let {
+            theirPublicKey = Base64.decode(it, Base64.NO_WRAP)
+        }
+        olmSAS = OlmSAS()
+        ourPublicKey = Base64.decode(olmSAS!!.publicKey, Base64.NO_WRAP)
+    }
+
+    override suspend fun connect(): String {
+        if (olmSAS == null) {
+            throw RuntimeException("Channel closed")
+        }
+        val isInitiator = theirPublicKey == null
+
+        if (isInitiator) {
+//            Timber.tag(TAG).i("Waiting for other device to send their public key")
+            val res = this.receiveAsPayload() ?: throw RuntimeException("No reply from other device")
+
+            if (res.key == null) {
+                throw RendezvousError(
+                        "Unsupported algorithm: ${res.algorithm}",
+                        RendezvousFailureReason.UnsupportedAlgorithm,
+                )
+            }
+            theirPublicKey = Base64.decode(res.key, Base64.NO_WRAP)
+        } else {
+            // send our public key unencrypted
+//            Timber.tag(TAG).i("Sending public key")
+            send(
+                    ECDHPayload(
+                    algorithm = SecureRendezvousChannelAlgorithm.ECDH_V1,
+                    key = Base64.encodeToString(ourPublicKey, Base64.NO_WRAP)
+            )
+            )
+        }
+
+        olmSAS!!.setTheirPublicKey(Base64.encodeToString(theirPublicKey, Base64.NO_WRAP))
```

Comment:
most of olm object are not thread safe. Would be more carefull here and use some mutex